com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ApplicationArchiveExtractor.java

```diff
@@ -0,0 +1,200 @@
+        do {
+            if (isFile(streamProducer.getStreamEntryName())) {
+                saveStreamToDirectory(streamProducer.getInputStream(), streamProducer.getStreamEntryName(), streamProducer.getFileName(),
+                    dirPath, callback);
+            }
+        } while (streamProducer.getNextInputStream() != null);
+        return dirPath;
+    }
+
+    private void saveEntries(InputStreamProducer streamProducer, Path dirPath, ExtractStatusCallback callback)
+        throws IOException, SLException {
+        ZipInputStream inputStream = (ZipInputStream) streamProducer.getInputStream();
+        for (ZipEntry entry; (entry = inputStream.getNextEntry()) != null;) {
+            if (!entry.getName()
+                .startsWith(streamProducer.getStreamEntryName())) {
+                continue;
+            }
+            validateEntry(entry.getName());
+            Path filePath = resolveTempEntryPath(entry.getName(), streamProducer.getStreamEntryName(), dirPath);
+            saveEntry(streamProducer.getInputStream(), entry, filePath, callback);
+        }
+    }
+
+    private void copy(InputStream input, OutputStream output, ExtractStatusCallback callback) throws IOException, SLException {
+        byte[] buffer = new byte[BUFFER_SIZE];
+        int numberOfReadBytes = 0;
+        while ((numberOfReadBytes = input.read(buffer)) != -1) {
+            callback.onBytesToWrite(numberOfReadBytes);
+            output.write(buffer, 0, numberOfReadBytes);
+        }
+    }
+
+    private void saveEntry(InputStream inputStream, ZipEntry entry, Path filePath, ExtractStatusCallback callback)
+        throws IOException, SLException {
+        if (!isFile(entry.getName())) {
+            Files.createDirectories(filePath);
+            return;
+        }
+
+        Files.createDirectories(filePath.getParent());
+        filePath = Files.createFile(filePath);
+        try (OutputStream outputStream = Files.newOutputStream(filePath)) {
+            copy(inputStream, outputStream, callback);
+        }
+    }
+
+    private void saveStreamToDirectory(InputStream inputStream, String entryName, String dirName, Path dirPath,
+        ExtractStatusCallback callback) throws IOException, SLException {
+        validateEntry(entryName);
+        Path filePath = resolveTempEntryPath(entryName, dirName, dirPath);
+        Files.createDirectories(filePath.getParent());
+        Files.createFile(filePath);
+        try (OutputStream outputStream = Files.newOutputStream(filePath)) {
+            copy(inputStream, outputStream, callback);
+        }
+    }
+
+    private boolean isFile(String fileName) {
+        return !fileName.endsWith(ARCHIVE_FILE_SEPARATOR);
+    }
+
+    private String getPrefix(String nameOfFile) {
+        int indexOfDot = nameOfFile.lastIndexOf('.');
+        String prefix = (indexOfDot > 0) ? nameOfFile.substring(0, indexOfDot) : nameOfFile;
+        if (prefix.length() < 3) {
+            prefix = prefix + "000";
```

Comment:
What is the purpose of this?

Reply:
It is required by File.createTempFile(prefix, suffix) that the prefix is at least 3 characters long otherwise it throws an exception.
This assignment wasn't implemented by me, it was just moved from another class.

Reply:
Okay.